core/transaction.go

```diff
@@ -229,102 +231,131 @@ func (l *L1HandlerTransaction) signature() []*felt.Felt {
 	switch t := transaction.(type) {
 	case *DeclareTransaction:
-		return declareTransactionHash(t, network)
+		return declareTransactionHash(t, n)
 	case *InvokeTransaction:
-		return invokeTransactionHash(t, network)
+		return invokeTransactionHash(t, n)
 	case *DeployTransaction:
-		return deployTransactionHash(t, network)
+		// deploy transactions are deprecated after re-genesis therefore we don't verify
+		// transaction hash
+		return nil, nil
 	case *L1HandlerTransaction:
-		return l1HandlerTransactionHash(t, network)
+		return l1HandlerTransactionHash(t, n)
 	case *DeployAccountTransaction:
-		return deployAccountTransactionHash(t, network)
+		return deployAccountTransactionHash(t, n)
 	default:
 		return nil, ErrUnknownTransaction
 	}
 }
 
-func deployTransactionHash(d *DeployTransaction, network utils.Network) (*felt.Felt, error) {
-	snKeccakConstructor, err := crypto.StarknetKeccak([]byte("constructor"))
-	if err != nil {
-		return nil, err
-	}
-	if d.Version.IsZero() || d.Version.IsOne() {
-		return crypto.PedersenArray(
-			new(felt.Felt).SetBytes([]byte("deploy")),
-			d.Version,
-			d.ContractAddress,
-			snKeccakConstructor,
-			crypto.PedersenArray(d.ConstructorCallData...),
-			new(felt.Felt),
-			network.ChainId(),
-		), nil
-	}
-	return nil, ErrInvalidTransactionVersion{d, d.Version.Text(10)}
-}
-
-func invokeTransactionHash(i *InvokeTransaction, network utils.Network) (*felt.Felt, error) {
-	invokeFelt := new(felt.Felt).SetBytes([]byte("invoke"))
+func invokeTransactionHash(i *InvokeTransaction, n utils.Network) (*felt.Felt, error) {
 	if i.Version.IsZero() {
-		return crypto.PedersenArray(
-			invokeFelt,
-			i.ContractAddress,
-			i.EntryPointSelector,
-			crypto.PedersenArray(i.CallData...),
-			network.ChainId(),
-		), nil
+		// Due to inconsistencies in version 0 hash calculation we don't verify the hash
+		return nil, nil
 	} else if i.Version.IsOne() {
 		return crypto.PedersenArray(
-			invokeFelt,
+			new(felt.Felt).SetBytes([]byte("invoke")),
 			i.Version,
 			i.ContractAddress,
 			new(felt.Felt),
 			crypto.PedersenArray(i.CallData...),
 			i.MaxFee,
-			network.ChainId(),
+			n.ChainId(),
 			i.Nonce,
 		), nil
 	}
 	return nil, ErrInvalidTransactionVersion{i, i.Version.Text(10)}
 }
 
-func declareTransactionHash(d *DeclareTransaction, network utils.Network) (*felt.Felt, error) {
-	declareFelt := new(felt.Felt).SetBytes([]byte("declare"))
+func declareTransactionHash(d *DeclareTransaction, n utils.Network) (*felt.Felt, error) {
 	if d.Version.IsZero() {
-		return crypto.PedersenArray(
-			declareFelt,
-			d.Version,
-			d.SenderAddress,
-			new(felt.Felt),
-			crypto.PedersenArray(make([]*felt.Felt, 0)...),
-			d.MaxFee,
-			network.ChainId(),
-			d.ClassHash,
-		), nil
+		// Due to inconsistencies in version 0 hash calculation we don't verify the hash
+		return nil, nil
 	} else if d.Version.IsOne() {
 		return crypto.PedersenArray(
-			declareFelt,
+			new(felt.Felt).SetBytes([]byte("declare")),
```

Comment:
can we make these constants global vars for each Tx type? could save use some calls to GC. 

We could hash them beforehand as well, so we avoid repeating the same calculations. But I am not sure if that is possible with the existing `PedersenArray` API we have

Reply:
We should implement an API that allows building a hash incrementally

Reply:
I will make the global vars. 

I am not sure how building incrementally will help. As we don't share hash inputs.

Reply:
Different types does not share inputs, but for a given type, first round of Pedersen always gives the same result since the strings are constant.

Reply:
We can have this task as a good first issue?

Reply:
Sounds good